File: MapandLoc/app/src/main/java/com/codewithamit/mapandloc/MapsActivity.java
package com.codewithamit.mapandloc;

import androidx.annotation.NonNull;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.FragmentActivity;

import android.Manifest;
import android.content.Context;
import android.content.pm.PackageManager;
import android.location.Address;
import android.location.Geocoder;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.widget.Toast;

import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;

import java.io.IOException;
import java.util.List;
import java.util.Locale;

public class MapsActivity extends FragmentActivity implements OnMapReadyCallback {

    LocationManager locationManager;
    LocationListener locationListener;

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);

        if (requestCode == 1) {
            if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {

                if(ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)== PackageManager.PERMISSION_GRANTED) {
                    locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 0, 0, locationListener);
                }
            }
        }
    }

    private GoogleMap mMap;

    @Override

    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_maps);
        // Obtain the SupportMapFragment and get notified when the map is ready to be used.
        SupportMapFragment mapFragment = (SupportMapFragment) getSupportFragmentManager()
                .findFragmentById(R.id.map);
        mapFragment.getMapAsync(this);
    }


    @Override
    public void onMapReady(GoogleMap googleMap) {
        mMap = googleMap;

        locationManager = (LocationManager) this.getSystemService(Context.LOCATION_SERVICE);
        locationListener = new LocationListener() {

            @Override
            public void onLocationChanged(@NonNull Location location) {
                // Add a marker in home and move the camera
                LatLng userLocation = new LatLng(location.getLatitude(), location.getLongitude());
                mMap.clear();
                mMap.addMarker(new MarkerOptions().position(userLocation).title("your location"));
                mMap.moveCamera(CameraUpdateFactory.newLatLng(userLocation));

                Geocoder geocoder = new Geocoder(getApplicationContext(), Locale.getDefault());

                try {
                    List<Address> listAddresses = geocoder.getFromLocation(location.getLatitude(),location.getLongitude(),1);

                    if(listAddresses!=null && listAddresses.size()>0){
                        Log.i("place info",listAddresses.get(0).toString());

                        String address = "";

                       if( listAddresses.get(0).getSubThoroughfare() !=null ){
                           address +=listAddresses.get(0).getSubThoroughfare() +"  ";
                       }
                       if(listAddresses.get(0).getThoroughfare() !=null){
                           address +=listAddresses.get(0).getThoroughfare()+",  ";
                       }
                        if(listAddresses.get(0).getFeatureName() !=null){
                            address +=listAddresses.get(0).getFeatureName()+",  ";
                        }
                        if(listAddresses.get(0).getLocality() !=null){
                            address +=listAddresses.get(0).getLocality()+",  ";
                        }
                        if(listAddresses.get(0).getPostalCode() !=null){
                            address +=listAddresses.get(0).getPostalCode()+",  ";
                        }
                        if(listAddresses.get(0).getCountryName() !=null){
                            address +=listAddresses.get(0).getCountryName();
                        }

                        Toast.makeText(MapsActivity.this,address,Toast.LENGTH_SHORT).show();
                    }

                } catch (IOException e) {
                    e.printStackTrace();
                }

            }
        };
        if (Build.VERSION.SDK_INT < 23) {

            if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {

                return;
            }
            locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER, 0, 0, locationListener);

        }else{
            if(ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)!= PackageManager.PERMISSION_GRANTED){

                ActivityCompat.requestPermissions(this,new String[]{Manifest.permission.ACCESS_FINE_LOCATION},1);
            }else{
                locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER,0,0,locationListener);

                Location lastKnownLocation = locationManager.getLastKnownLocation(LocationManager.GPS_PROVIDER);

                LatLng userLocation = new LatLng(lastKnownLocation.getLatitude(), lastKnownLocation.getLongitude());
                mMap.clear();
                mMap.addMarker(new MarkerOptions().position(userLocation).title("your location"));
                mMap.moveCamera(CameraUpdateFactory.newLatLng(userLocation));

            }
        }


    }
}

File: MapsAndLocation/app/src/main/java/com/example/mapsandlocation/MapsActivity.kt
package com.example.mapsandlocation

import android.Manifest
import android.annotation.SuppressLint
import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationManager
import android.os.Build
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Looper
import android.provider.Settings
import android.widget.Toast
import androidx.annotation.RequiresApi
import androidx.appcompat.app.AlertDialog
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat

import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import com.example.mapsandlocation.databinding.ActivityMapsBinding
import com.google.android.gms.location.LocationCallback
import com.google.android.gms.location.LocationRequest
import com.google.android.gms.location.LocationResult
import com.google.android.gms.location.LocationServices
import com.google.android.gms.maps.model.PolylineOptions

class MapsActivity : AppCompatActivity(), OnMapReadyCallback {

    private lateinit var mMap: GoogleMap
    private lateinit var binding: ActivityMapsBinding

    val locationManager by lazy {
        getSystemService(Context.LOCATION_SERVICE) as LocationManager
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityMapsBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val mapFragment = supportFragmentManager
                .findFragmentById(R.id.map) as SupportMapFragment
        mapFragment.getMapAsync(this)
    }

    @RequiresApi(Build.VERSION_CODES.M)
    override fun onStart() {
        requestAccessFineLocation()
        super.onStart()
        when{
            isFineLocationGranted()->{
                when{
                    isLocationEnable()->setupLocationListener()
                    else->showGPSNotEnableDialog()
                }
            }
            else->requestAccessFineLocation()
        }
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        when(requestCode){
            999 -> if(grantResults.isNotEmpty()&&grantResults[0]==PackageManager.PERMISSION_GRANTED){
//                setupLocationListener()
                when{
                    isLocationEnable()->setupLocationListener()
                    else->showGPSNotEnableDialog()
                }
            }else{
                Toast.makeText(this,"permission not granted",Toast.LENGTH_LONG).show()
            }
        }
    }

    private fun setupLocationListener() {
        val fusedLocationProviderClient = LocationServices.getFusedLocationProviderClient(this)

        val locationRequest = LocationRequest()
            .setInterval(2000)
            .setFastestInterval(2000)
            .setSmallestDisplacement(1f)
            .setPriority(LocationRequest.PRIORITY_HIGH_ACCURACY)

        if (ActivityCompat.checkSelfPermission(
                this,
                Manifest.permission.ACCESS_FINE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(
                this,
                Manifest.permission.ACCESS_COARSE_LOCATION
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            return
        }
        fusedLocationProviderClient.requestLocationUpdates(locationRequest,object :LocationCallback(){
            override fun onLocationResult(locationResult: LocationResult) {
                super.onLocationResult(locationResult)
                for (location in locationResult.locations){
                    val current = LatLng(location.latitude, location.longitude)

                    if(::mMap.isInitialized){
                        mMap.addMarker(MarkerOptions().position(current).title("My Home"))
                        mMap.moveCamera(CameraUpdateFactory.newLatLng(current))
                        }
                }
            }
        }, Looper.myLooper()
        )

    }

//    @SuppressLint("missing permission")
//    private fun setupLocationListener() {
//
//
//        val providers = locationManager.getProviders(true)
//
//        var l:Location? = null
//        for (i in providers.indices.reversed()){
//            if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
//                return
//            }
//            l= locationManager.getLastKnownLocation(providers[i])
//            if (l != null)break
//        }
//
//        l?.let {
//            if(::mMap.isInitialized){
//                val current = LatLng(15.466570, 73.852734)
//                mMap.addMarker(MarkerOptions().position(current).title("My Home"))
//                mMap.moveCamera(CameraUpdateFactory.newLatLng(current))
//            }
//        }
//
//    }

    @RequiresApi(Build.VERSION_CODES.M)
    fun isFineLocationGranted():Boolean{
        return checkSelfPermission(Manifest.permission.ACCESS_FINE_LOCATION) ==PackageManager.PERMISSION_GRANTED
    }

    @RequiresApi(Build.VERSION_CODES.M)
    private fun requestAccessFineLocation() {
        this.requestPermissions(
                arrayOf(Manifest.permission.ACCESS_FINE_LOCATION),999
        )
    }

    fun isLocationEnable():Boolean {
        return locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER) ||
                locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER)
    }

    fun showGPSNotEnableDialog(){
        AlertDialog.Builder(this)
                .setTitle("Enable GPS")
                .setMessage("GPS is required")
                .setCancelable(false)
                .setPositiveButton("Enable Now"){ dialogInterface: DialogInterface, i: Int ->
                    startActivity(Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS))
                    dialogInterface.dismiss()
                }.show()
    }

    override fun onMapReady(googleMap: GoogleMap) {
        mMap = googleMap

        mMap.uiSettings.apply {
            isZoomControlsEnabled=true
            isZoomGesturesEnabled=true
            isMyLocationButtonEnabled=true
            isCompassEnabled=true
        }

        mMap.setMaxZoomPreference(22f)

        val sydney = LatLng(15.466570, 73.852734)
        mMap.addMarker(MarkerOptions().position(sydney).title("My Home"))
        mMap.moveCamera(CameraUpdateFactory.newLatLng(sydney))

        mMap.addPolyline(
                PolylineOptions().add(LatLng(23.45,75.34), LatLng(45.45,95.34))
                        .color(ContextCompat.getColor(baseContext,R.color.design_default_color_primary))
        ).width=2f
    }
}